src/analytics-pre-init.ts

```diff
@@ -0,0 +1,279 @@
+
+/**
+ * A list of the method calls before initialization for snippet users
+ * For example, [["track", "foo", {bar: 123}], ["page"], ["on", "ready", function(){..}]
+ */
+type SnippetBuffer = SnippetWindowBufferedMethodCall[]
+
+/**
+ * Fetch the buffered method calls from the window object and normalize them.
+ */
+export const getSnippetWindowBuffer = (): PreInitMethodCall[] => {
+  const wa = window.analytics
+  const buffered =
+    // @ts-expect-error
+    (wa && wa[0] ? [...wa] : []) as SnippetBuffer
+  return normalizeSnippetBuffer(buffered)
+}
+
+/**
+ * Infer return type; if return type is promise, unwrap it.
+ */
+type ReturnTypeUnwrap<Fn> = Fn extends (...args: any[]) => infer ReturnT
+  ? ReturnT extends PromiseLike<infer Unwrapped>
+    ? Unwrapped
+    : ReturnT
+  : never
+
+type MethodCallMap = Partial<Record<PreInitMethodName, PreInitMethodCall[]>>
+
+/**
+ *  Represents any and all the buffered method calls that occurred before initialization.
+ */
+export class PreInitMethodCallBuffer {
+  private _value = {} as MethodCallMap
+
+  public toArray(): PreInitMethodCall[] {
+    return Object.values(this._value).reduce((acc, v) => {
+      return acc.concat(...v)
+    }, [] as PreInitMethodCall[])
+  }
+
+  public getCalls(methodName: PreInitMethodName): PreInitMethodCall[] {
+    return this._value[methodName] || []
+  }
+
+  push(...calls: PreInitMethodCall[]): void {
+    calls.forEach((el) => {
+      if (this._value[el.method]) {
+        this._value[el.method]?.push(el)
+      } else {
+        this._value[el.method] = [el]
+      }
+    })
+  }
+
+  clear(): void {
+    this._value = {} as MethodCallMap
+  }
+}
+
+/**
+ *  Call method and mark as "called"
+ *  This function should never throw an error
+ */
+export async function callAnalyticsMethod<T extends PreInitMethodName>(
+  analytics: Analytics,
+  methodCall: PreInitMethodCall<T>
+): Promise<void> {
+  const { method, args, resolve, reject } = methodCall
+  try {
+    if (methodCall.called) {
+      return undefined
+    }
+
+    methodCall.called = true
+
+    const result = await (analytics[method] as Function)(...args)
+    resolve(result)
+  } catch (err) {
+    reject(err)
+  }
+}
+
+type AnalyticsLoader = (
+  preInitBuffer: PreInitMethodCallBuffer
+) => Promise<[Analytics, Context]>
+
+export class AnalyticsBuffered implements PromiseLike<[Analytics, Context]> {
+  public instance?: Analytics
+  public ctx?: Context
+  private preInitBuffer = new PreInitMethodCallBuffer()
+  private promise: Promise<[Analytics, Context]>
+  constructor(loader: AnalyticsLoader) {
+    this.promise = loader(this.preInitBuffer)
+    this.promise
+      .then(([ajs, ctx]) => {
+        this.instance = ajs
+        this.ctx = ctx
+      })
+      .catch(() => {
+        // intentionally do nothing...
+        // this result of this promise will be caught by the 'catch' block on this class.
+      })
+  }
+
+  private _createMethod<T extends PreInitMethodName>(methodName: T) {
```

Comment:
Thanks, I've been thinking this over
> on/off/once 
this is a good call out -- I think this is in the spirit of the event emitter interface. Since we're registering a handler, and a user knowing _when_ the handler is registered is of limited use -- it's worth the trade off.

> setAnonymousId 
I think we just document this behavior, we don't actually have any way of returning anything but a promise<ID | undefined> 

> reset
as is -- Promise<void> is fine.

> debug 
as is -- Promise<void>